<a href="https://colab.research.google.com/github/princealexanderm/ICT_DAS_Prince_Alex/blob/main/Assignment_on_Model_Selection_Prince_12_Aug_24.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Name: Prince Alex**<br>
**Topic: Model Selection**<br>
**Date: 12/08/24**

Do the following on titanic dataset
1. Load the dataset into python environment
2. Do all the necessary pre-processing steps
3. Create kNN and SVM models
4. Do k-fold and stratified stratified k-fold cross validation techniques and find the average accuracy score of the models

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
df=pd.read_csv('/content/drive/MyDrive/ICT/titanic_dataset .csv')
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [ ]:
df.shape

(891, 12)

In [ ]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

# Checking Null values

In [ ]:
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [ ]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [ ]:
df.nunique()

PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             88
SibSp            7
Parch            7
Ticket         681
Fare           248
Cabin          147
Embarked         3
dtype: int64

In [ ]:
df.shape

(891, 12)

Here 'cabin' is an unnecessary data. Let's drop it

In [ ]:
df.drop(['Cabin','Name','Ticket'],axis=1,inplace=True)

In [ ]:
df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Sex              0
Age            177
SibSp            0
Parch            0
Fare             0
Embarked         2
dtype: int64

# Filling null values

Null values in 'Age' and 'Embarked'. Lets fill Age with mean and embarked with mode

In [ ]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch',
       'Fare', 'Embarked'],
      dtype='object')

In [ ]:
df['Age'].fillna(df['Age'].mean(),inplace=True)
df['Embarked'].fillna(df['Embarked'].mode()[0],inplace=True)

In [ ]:
df.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Sex            0
Age            0
SibSp          0
Parch          0
Fare           0
Embarked       0
dtype: int64

All null values are filled

#Encoding using one-hot-encoder

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Sex          891 non-null    object 
 4   Age          891 non-null    float64
 5   SibSp        891 non-null    int64  
 6   Parch        891 non-null    int64  
 7   Fare         891 non-null    float64
 8   Embarked     891 non-null    object 
dtypes: float64(2), int64(5), object(2)
memory usage: 62.8+ KB


sex, embarked has to be encoded

In [ ]:
from sklearn.preprocessing import StandardScaler , MinMaxScaler , LabelEncoder, OneHotEncoder
categorical_columns = ['Sex','Embarked']
one_hot_enc = OneHotEncoder(sparse=False,drop='first')
encoded_data = one_hot_enc.fit_transform(df[categorical_columns])

encoded_data

/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:975: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


array([[1., 0., 1.],
       [0., 0., 0.],
       [0., 0., 1.],
       ...,
       [0., 0., 1.],
       [1., 0., 0.],
       [1., 1., 0.]])

In [ ]:
encoded_categorical_df = pd.DataFrame(encoded_data,
                                      columns=one_hot_enc.get_feature_names_out(categorical_columns))
encoded_categorical_df.head()

Sex_male  Embarked_Q  Embarked_S
0       1.0         0.0         1.0
1       0.0         0.0         0.0
2       0.0         0.0         1.0
3       0.0         0.0         1.0
4       1.0         0.0         1.0

In [ ]:
data_encoded = pd.concat([df.drop(columns=categorical_columns),encoded_categorical_df],axis=1)
data_encoded.head()

PassengerId  Survived  Pclass   Age  SibSp  Parch     Fare  Sex_male  \
0            1         0       3  22.0      1      0   7.2500       1.0   
1            2         1       1  38.0      1      0  71.2833       0.0   
2            3         1       3  26.0      0      0   7.9250       0.0   
3            4         1       1  35.0      1      0  53.1000       0.0   
4            5         0       3  35.0      0      0   8.0500       1.0   

   Embarked_Q  Embarked_S  
0         0.0         1.0  
1         0.0         0.0  
2         0.0         1.0  
3         0.0         1.0  
4         0.0         1.0

#Scaling

In [ ]:
numerical_columns = ['Age','Fare']

std_scaler = StandardScaler()

data_std_scaled = data_encoded.copy()

data_std_scaled[numerical_columns] = std_scaler.fit_transform(data_std_scaled[numerical_columns])
data_std_scaled.head()

PassengerId  Survived  Pclass       Age  SibSp  Parch      Fare  Sex_male  \
0            1         0       3 -0.592481      1      0 -0.502445       1.0   
1            2         1       1  0.638789      1      0  0.786845       0.0   
2            3         1       3 -0.284663      0      0 -0.488854       0.0   
3            4         1       1  0.407926      1      0  0.420730       0.0   
4            5         0       3  0.407926      0      0 -0.486337       1.0   

   Embarked_Q  Embarked_S  
0         0.0         1.0  
1         0.0         0.0  
2         0.0         1.0  
3         0.0         1.0  
4         0.0         1.0

In [ ]:
data_std_scaled.describe()

PassengerId    Survived      Pclass           Age       SibSp  \
count   891.000000  891.000000  891.000000  8.910000e+02  891.000000   
mean    446.000000    0.383838    2.308642  2.232906e-16    0.523008   
std     257.353842    0.486592    0.836071  1.000562e+00    1.102743   
min       1.000000    0.000000    1.000000 -2.253155e+00    0.000000   
25%     223.500000    0.000000    2.000000 -5.924806e-01    0.000000   
50%     446.000000    0.000000    3.000000  0.000000e+00    0.000000   
75%     668.500000    1.000000    3.000000  4.079260e-01    1.000000   
max     891.000000    1.000000    3.000000  3.870872e+00    8.000000   

            Parch          Fare    Sex_male  Embarked_Q  Embarked_S  
count  891.000000  8.910000e+02  891.000000  891.000000  891.000000  
mean     0.381594  3.987333e-18    0.647587    0.086420    0.725028  
std      0.806057  1.000562e+00    0.477990    0.281141    0.446751  
min      0.000000 -6.484217e-01    0.000000    0.000000    0.000000  
25%      0.000000 -4.891482e-01    0.000000    0.000000    0.000000  
50%      0.000000 -3.573909e-01    1.000000    0.000000    1.000000  
75%      0.000000 -2.424635e-02    1.000000    0.000000    1.000000  
max      6.000000  9.667167e+00    1.000000    1.000000    1.000000

In [ ]:
data_encoded.drop(columns=['PassengerId'],inplace=True)

In [ ]:
df=data_encoded

In [ ]:
df.head()

Survived  Pclass   Age  SibSp  Parch     Fare  Sex_male  Embarked_Q  \
0         0       3  22.0      1      0   7.2500       1.0         0.0   
1         1       1  38.0      1      0  71.2833       0.0         0.0   
2         1       3  26.0      0      0   7.9250       0.0         0.0   
3         1       1  35.0      1      0  53.1000       0.0         0.0   
4         0       3  35.0      0      0   8.0500       1.0         0.0   

   Embarked_S  
0         1.0  
1         0.0  
2         1.0  
3         1.0  
4         1.0

#Preprocessing completede

---

#split the data as target and independet variabe

In [ ]:
X=df.drop('Survived',axis=1)
y=df['Survived']

#Split the data for Xtrain and ytrain

In [ ]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)

---

#KNN

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

knn_model=KNeighborsClassifier()
knn_model.fit(X_train,y_train)

KNeighborsClassifier()

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score
y_pred=knn_model.predict(X_test)

print('Accuracy is', accuracy_score(y_test, y_pred))
print('Precision is', precision_score(y_test, y_pred))
print('Recall is', recall_score(y_test, y_pred))
print('f1 score is', f1_score(y_test,y_pred))

Accuracy is 0.6983240223463687
Precision is 0.6666666666666666
Recall is 0.5405405405405406
f1 score is 0.5970149253731343


---

#SVM

In [ ]:
from sklearn.svm import SVC

svc_model=SVC()
svc_model.fit(X_train,y_train)

SVC()

In [ ]:
from sklearn.metrics import accuracy_score
y_pred=svc_model.predict(X_test)

print('Accuracy is', accuracy_score(y_test, y_pred))
print('Precision is', precision_score(y_test, y_pred))
print('Recall is', recall_score(y_test, y_pred))
print('f1 score is', f1_score(y_test,y_pred))

Accuracy is 0.659217877094972
Precision is 0.76
Recall is 0.25675675675675674
f1 score is 0.3838383838383838


---

#K-fold Cross validation-KNN model

In [ ]:
from sklearn.model_selection import KFold

In [ ]:
kfold_val=KFold(10) #default it is 5, if you don't give argument

In [ ]:
for train_index,test_index in kfold_val.split(X,y):
  print('Training_index:',train_index)
  print('Test Index:',test_index)

Training_index: [ 90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233
 234 235 236 237 238 239 240 241 242 243 244 245 246 247 248 249 250 251
 252 253 254 255 256 257 258 259 260 261 262 263 264 265 266 267 268 269
 270 271 272 273 274 275 276 277 278 279 280 281 282 283 284 285 286 287
 288 289 290 291 292 293 294 295 296 297 298 299 300 301 302 303 304 305
 306 307 308 309 310 311 312 313 314 315 316 317 318 319 320 321 322 323
 324 325 326 327 328 329 330 331 33

#Cross validation

In [ ]:
from sklearn.model_selection import cross_val_score
kv_result=cross_val_score(knn_model,X,y,cv=kfold_val)

In [ ]:
kv_result

array([0.62222222, 0.66292135, 0.70786517, 0.65168539, 0.70786517,
       0.73033708, 0.69662921, 0.73033708, 0.6741573 , 0.78651685])

In [ ]:
mean_result_knn=kv_result.mean()

In [ ]:
mean_result_knn

0.6970536828963796

#Stratified cross validationn-KNN model

In [ ]:
from sklearn.model_selection import StratifiedKFold
sk_fold=StratifiedKFold(10)

In [ ]:
skv_result=cross_val_score(knn_model,X,y,cv=sk_fold)

In [ ]:
skv_result

array([0.62222222, 0.65168539, 0.68539326, 0.69662921, 0.73033708,
       0.70786517, 0.73033708, 0.71910112, 0.65168539, 0.7752809 ])

In [ ]:
skv_result_knn=skv_result.mean()

In [ ]:
skv_result_knn

0.6970536828963796

---

#K-fold Cross validation-SVM model

In [ ]:
from sklearn.model_selection import KFold
kfold_val=KFold(10) #default it is 5, if you don't give argument
for train_index,test_index in kfold_val.split(X,y):
  print('Training_index:',train_index)
  print('Test Index:',test_index)

Training_index: [ 90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233
 234 235 236 237 238 239 240 241 242 243 244 245 246 247 248 249 250 251
 252 253 254 255 256 257 258 259 260 261 262 263 264 265 266 267 268 269
 270 271 272 273 274 275 276 277 278 279 280 281 282 283 284 285 286 287
 288 289 290 291 292 293 294 295 296 297 298 299 300 301 302 303 304 305
 306 307 308 309 310 311 312 313 314 315 316 317 318 319 320 321 322 323
 324 325 326 327 328 329 330 331 33

#Cross validation

In [ ]:
from sklearn.model_selection import cross_val_score
kv_result_svc=cross_val_score(svc_model,X,y,cv=kfold_val)

In [ ]:
kv_result_svc

array([0.6       , 0.6741573 , 0.66292135, 0.6741573 , 0.65168539,
       0.69662921, 0.65168539, 0.74157303, 0.73033708, 0.69662921])

In [ ]:
mean_result_svc=kv_result_svc.mean()

In [ ]:
mean_result_svc

0.6779775280898878

#Stratified cross validationn-SVM model

In [ ]:
from sklearn.model_selection import StratifiedKFold
sk_fold=StratifiedKFold(10)

In [ ]:
skv_result_svc=cross_val_score(svc_model,X,y,cv=sk_fold)

In [ ]:
skv_result_svc

array([0.62222222, 0.58426966, 0.68539326, 0.74157303, 0.69662921,
       0.69662921, 0.68539326, 0.71910112, 0.70786517, 0.6741573 ])

In [ ]:
skv_result_svc2=skv_result_svc.mean()

In [ ]:
skv_result_svc2

0.6813233458177278

---

#Accuracy score

1. KNN-K-fold Cross validation:**0.6970536828963796**
2. KNN-Stratified cross validationn:**0.6970536828963796**
3. SVM-K-fold Cross validation:**0.6779775280898878**
4. SVM-Stratified cross validationn:**0.6813233458177278**